File: src/main/java/com/myudog/myulib/internal/generator/TorusGenerator.java
package com.myudog.myulib.internal.generator;
import com.myudog.myulib.api.shape.IShape;
import net.minecraft.world.phys.Vec3;import java.util.ArrayList;
import java.util.List;
public final class TorusGenerator implements IShape {
    @Override
    public List<Vec3> getOutlinePoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double majorR = size.x;
        double minorR = size.y;
        int majorCount = Math.max(12, (int) Math.round(2.0 * Math.PI * majorR * Math.max(1.0, density)));
        int minorCount = Math.max(8, (int) Math.round(2.0 * Math.PI * minorR * Math.max(1.0, density)));
        for (int i = 0; i < majorCount; i++) {
            double theta = 2.0 * Math.PI * i / majorCount;
            for (int j = 0; j < minorCount; j++) {
                double phi = 2.0 * Math.PI * j / minorCount;
                double x = (majorR + minorR * Math.cos(phi)) * Math.cos(theta);
                double y = minorR * Math.sin(phi);
                double z = (majorR + minorR * Math.cos(phi)) * Math.sin(theta);
                points.add(new Vec3(x, y, z));
            }
        }
        return points;
    }
    @Override
    public List<Vec3> getSolidPoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double r = 0.1;
        double rStep = 1.0 / Math.max(0.1, density);
        while (r <= size.y) {
            points.addAll(getOutlinePoints(new Vec3(size.x, r, 0.0), density));
            r += rStep;
        }
        return points;
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/generator/TorusGenerator.kt
package com.myudog.myulib.internal.generator

import com.myudog.myulib.api.shape.IShape
import net.minecraft.util.math.Vec3d
import kotlin.math.*

internal class TorusGenerator : IShape {
    override fun getOutlinePoints(size: Vec3d, density: Double): List<Vec3d> {
        val points = mutableListOf<Vec3d>()
        val majorR = size.x // 主半徑
        val minorR = size.y // 管半徑

        // 需要兩個角度來定義甜甜圈上的點
        val majorCount = (2 * PI * majorR * density).toInt()
        val minorCount = (2 * PI * minorR * density).toInt()

        for (i in 0 until majorCount) {
            val theta = 2 * PI * i / majorCount
            for (j in 0 until minorCount) {
                val phi = 2 * PI * j / minorCount

                // 甜甜圈參數方程
                val x = (majorR + minorR * cos(phi)) * cos(theta)
                val y = minorR * sin(phi)
                val z = (majorR + minorR * cos(phi)) * sin(theta)

                points.add(Vec3d(x, y, z))
            }
        }
        return points
    }

    override fun getSolidPoints(size: Vec3d, density: Double): List<Vec3d> {
        // 填充甜甜圈實心（透過多層管半徑）
        val points = mutableListOf<Vec3d>()
        var r = 0.1
        val rStep = 1.0 / density
        while (r <= size.y) { // 從中心到 Minor Radius
            points.addAll(getOutlinePoints(Vec3d(size.x, r, 0.0), density))
            r += rStep
        }
        return points
    }
}